### PR TITLE
Rework schema errors

### PR DESCRIPTION
**Pull request by lafrech**
I tried to address issues raised in https://github.com/marshmallow-code/marshmallow/pull/857#issuecomment-400737110 by refactoring the code.

The first commit (https://github.com/marshmallow-code/marshmallow/pull/862/commits/9f84247bad8693ec8e357252aa35d9b4e07093a8) changes the error structure in a corner case, but in a good way. It could be considered a bugfix. I can send a separate PR for this one if needed.

The other commits are about factorizing the code that stores schema errors by adapting `store_error` (that normally stores field errors) while it is currently duplicated in `Unmarshaller`. I'm happy with this rework (not "this is so perfect" happy, but rather "this is much better" happy), except one test is broken.

When `store_error` receives a `dict` as message, it does not merge it, nor does it append it to the current error messages list. It just uses it as error message.

```python
        if isinstance(messages, dict):
            errors[field_name] = messages
```

Hence the test I had to modify:

```python
-        assert errors['_schema'][0] == {'code': 'invalid_field'}
+        assert errors['_schema'] == {'code': 'invalid_field'}
```

This differs from the behaviour of the code duplicated in `Unmarshaller`. And changing this in `store_error` would break a lot of field errors tests.

While working on this, I ended up facing the issues described in https://github.com/marshmallow-code/marshmallow/issues/441. I think I'll either follow-up there or create a separate issue.

### CODE REVIEW

**Comment by lafrech** [comment]
Rebased.

**Comment by sloria** [comment]
I like the refactoring done here; the addition of `store_validation_error` makes sense.

Re: the changed test: I *think* it's an OK change. There is a valid use case for using dictionaries (rather than strings) as individual error messages. marshmallow-jsonapi, for example, [does this](https://github.com/marshmallow-code/marshmallow-jsonapi/blob/289b4ab750e2d5781e50b4f8cc0c82e5ebfefe9b/marshmallow_jsonapi/schema.py#L147-L154). But this use case is still supported by this PR, since the user can pass a list of dictionaries to `ValidationError`. Also, I think you're right to point out that this change could set up #441 nicely.

Because this is a breaking change, it'd be good to have one more set of eyes on this.

**Comment by lafrech** [comment]
Thanks for the feedback.

This PR does not really address #441. It is just a small step towards refactoring the error structure. Maybe it will help by centralizing the problem in one single piece of code. Regarding #441, I'd like to list all cases (different types like string, list, dict,... and which fields returns what) to get a clearer view and decide what should be valid or not. But that can be in a further step.

File: marshmallow/marshalling.py
Comment (deckar01) [inline marshmallow/marshalling.py:51]: > When store_error receives a dict as message, it does not merge it ... This differs from the behaviour of the code duplicated in Unmarshaller. And changing this in store_error would break a lot of field errors tests.

It looks like fields were originally ([8340fca5][0]) allowed to append dicts to the error list, but in a seemingly unrelated bug fix ([4698fe51][1]) it was regressed.

It's probably fine to overwrite some schema errors since field errors have been overwritten for so long, but it might be worth investigating what errors this is hiding. When the field error tests fail, is it because there are extra errors that the tests don't expect, or does it violate core expectations about the structure of the error? Are the hidden errors important?

[0]: https://github.com/marshmallow-code/marshmallow/commit/8340fca5868346e2e20d721d87bbc892354e6d19#diff-c4881f63f8b05b8d3eadf046c3dc7912R110
[1]: https://github.com/marshmallow-code/marshmallow/commit/4698fe51e6d756dca92912a9097b11a7df8a57f2#diff-c4881f63f8b05b8d3eadf046c3dc7912R111

**Comment by lafrech** [comment]
For the record, while working on the error structure (https://github.com/marshmallow-code/marshmallow/issues/441), I noticed that this PR fixes (what I think is) a corner case issue.

I modified a test in `test_deserialization` as a quick way of testing the output when errors are raised from different validators:

```python
    def test_nested_single_deserialization_to_dict(self):
        class SimpleBlogSerializer(Schema):
            title = fields.String()
            author = fields.Nested(SimpleUserSchema, required=True)
            @validates('author')
            def validate_author(self, value):
               if value != 42:
                   raise ValidationError('The answer to life the universe and everything.')
            @validates_schema(skip_on_field_errors=False)
            def validate_schema(self, data):
                raise ValidationError('This will never work', 'author')

        blog_dict = {
            'title': 'Gimme Shelter',
            'author': {'name': 'Mick', 'age': 'lol', 'email': 'mick@stones.com'},
        }
        result = SimpleBlogSerializer().load(blog_dict)
```

Without this PR, the error returned is 

`ValidationError: {'author': {'_field': ['The answer to life the universe and everything.'], 'age': ['Not a valid number.'], '_schema': ['This will never work']}}`

which is wrong: there shouldn't be `_schema`, here.

With this change, I get

`ValidationError: {'author': {'age': ['Not a valid number.'], '_field': ['The answer to life the universe and everything.', 'This will never work']}}`

which is the expected output.

**Comment by sloria** [comment]
@deckar01 Do you see any blockers to merging this?

**Comment by deckar01** [comment]
Feel free to ignore https://github.com/marshmallow-code/marshmallow/pull/862#discussion_r199907288. It is similar to existing behavior and it doesn't seem to be bothering anyone. Looks good to me otherwise.

**Review by deckar01 — APPROVED** [review]

**Comment by sloria** [comment]
Thanks @deckar01. @lafrech, care to update the changelog and upgrading guide?

**Comment by lafrech** [comment]
I rebased and updated the changelog. @sloria, do you see anything to add to the upgrading guide?

I still don't like the way `store_error` works, and the breaking change underlined by @deckar01 makes it slightly worse, but it is not that bad as long as we get this sorted for 3.0.

I think this should be dealt with in a wider reflection and I hope https://github.com/marshmallow-code/marshmallow/issues/879 will help clarifying things.

File: CHANGELOG.rst
Comment (sloria) [inline CHANGELOG.rst]: Typos: "schema-level"; "is" -> "are"

**Comment by sloria** [comment]
It would be good to document the changed behavior with `raise ValidationError({...})` in the upgrading guide. Certainly not a blocker to merging this if you don't have time right now, though.

File: CHANGELOG.rst
Comment (lafrech) [reply CHANGELOG.rst]: Oh, sorry. Fixed.

**Comment by lafrech** [comment]
Yeah, time and not sure how to formulate this.

I'd better leave this to you when you get the time.

**Pull request merged**
